dm/pkg/conn/basedb.go

```diff
@@ -175,6 +177,61 @@ func (d *BaseDB) GetBaseConn(ctx context.Context) (*BaseConn, error) {
 	return baseConn, nil
 }
 
+func (d *BaseDB) ExecContext(tctx *tcontext.Context, query string, args ...interface{}) (sql.Result, error) {
+	if tctx.L().Core().Enabled(zap.DebugLevel) {
+		tctx.L().Debug("exec context",
+			zap.String("query", utils.TruncateString(query, -1)),
+			zap.String("argument", utils.TruncateInterface(args, -1)))
+	}
+	return d.DB.ExecContext(tctx.Ctx, query, args...)
+}
+
+func (d *BaseDB) QueryContext(tctx *tcontext.Context, query string, args ...interface{}) (*sql.Rows, error) {
+	if tctx.L().Core().Enabled(zap.DebugLevel) {
+		tctx.L().Debug("query context",
+			zap.String("query", utils.TruncateString(query, -1)),
+			zap.String("argument", utils.TruncateInterface(args, -1)))
+	}
+	return d.DB.QueryContext(tctx.Ctx, query, args...)
+}
+
+func (d *BaseDB) DoTxWithRetry(tctx *tcontext.Context, queries []string, args [][]interface{}, retryer retry.Retryer) error {
```

Comment:
maybe we can wrap all db/conns to `SQLWithRetry`, I see lots of same logic in many repos 🤣

https://github.com/pingcap/tidb/blob/f88dab92f9c5ca3f2225a1dab19ff9b4e6aaaaa1/br/pkg/lightning/common/util.go#L141